File: packages/network/src/v2/schemas/decodeField.ts
import { DynamicSchemaType, StaticSchemaType } from "@latticexyz/schema-type/deprecated";
import { hexToArray } from "@latticexyz/utils";
import { Schema } from "../common";
import { decodeStaticField } from "./decodeStaticField";
import { decodeDynamicField } from "./decodeDynamicField";

export const decodeField = (schema: Schema, schemaIndex: number, hexData: string): Record<number, any> => {
  const data: Record<number, any> = {};
  const bytes = hexToArray(hexData);

  schema.staticFields.forEach((fieldType, index) => {
    if (index === schemaIndex) {
      data[schemaIndex] = decodeStaticField(fieldType as StaticSchemaType, bytes, 0);
    }
  });

  if (schema.dynamicFields.length > 0) {
    schema.dynamicFields.forEach((fieldType, i) => {
      const index = schema.staticFields.length + i;
      if (index === schemaIndex) {
        data[schemaIndex] = decodeDynamicField(fieldType as DynamicSchemaType, bytes);
      }
    });
  }

  return data;
};

File: packages/services/pkg/protocol-parser/decodeField.go
package protocolparser

import schematype "latticexyz/mud/packages/services/pkg/schema-type"

func DecodeField(schemaType schematype.SchemaType, data string) interface{} {
	if schemaType.IsDynamic() {
		return DecodeDynamicField(schemaType, data)
	} else {
		return DecodeStaticField(schemaType, data)
	}
}
